### PR TITLE
feat: Add cluster-config ConfigMap support to KonfluxInfo

### PR DESCRIPTION
**Pull request by gbenhaim**
### **User description**
Add support for cluster-wide configuration that can be consumed by PipelineRuns via a stable ConfigMap API

The configuration is stored in a "cluster-config" ConfigMap in the "konflux-info" namespace, readable by all authenticated users. User-provided values take precedence over auto-discovered values, enabling operators to override defaults while allowing automatic detection of cluster-specific settings.

Assisted-By: Cursor

___

### **PR Type**
Enhancement

___

### **Description**
- Add `ClusterConfig` API type to support cluster-wide key-value configuration

- Implement `cluster-config` ConfigMap reconciliation with user-provided and auto-discovered values

- Add `ClusterConfigDiscoverer` interface for extensible cluster configuration discovery

- Include comprehensive unit and integration tests for cluster configuration functionality

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["KonfluxInfo CR"] -->|"ClusterConfig spec"| B["ClusterConfigData struct"]
  B -->|"User-provided values"| C["MergeOver method"]
  D["ClusterConfigDiscoverer"] -->|"Auto-discovered values"| C
  C -->|"Merged config"| E["cluster-config ConfigMap"]
  E -->|"Readable by PipelineRuns"| F["Authenticated users"]
```

<details><summary><h3>File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>konfluxinfo_types.go</strong><dd><code>Add ClusterConfig types and merge logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/api/v1alpha1/konfluxinfo_types.go

<ul><li>Add <code>ClusterConfig</code> field to <code>KonfluxInfoSpec</code> for cluster-wide <br>configuration<br> <li> Define <code>ClusterConfigData</code> struct with 7 configuration fields (OIDC, <br>Fulcio, Rekor, TUF URLs)<br> <li> Implement <code>All()</code> iterator method to convert struct to map using Go 1.22 <br>range-over-func<br> <li> Implement <code>MergeOver()</code> method to merge discovered and user-provided <br>values with precedence<br> <li> Add <code>GetClusterConfigData()</code> accessor method for safe nil handling</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-975033ee15cf428afbbc8d14b37b6def2d08a01614d699e82b2bac2d70f5cf11">+123/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>konfluxinfo_controller.go</strong><dd><code>Implement cluster-config reconciliation logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/internal/controller/info/konfluxinfo_controller.go

<ul><li>Add <code>clusterConfigMapName</code> constant for <code>cluster-config</code> ConfigMap<br> <li> Define 7 ConfigMap key constants with breaking change warnings<br> <li> Add <code>ClusterConfigDiscoverer</code> interface for dependency injection<br> <li> Add <code>defaultClusterConfigDiscoverer</code> implementation returning empty <br>values<br> <li> Add <code>NewKonfluxInfoReconciler()</code> constructor to initialize discoverer<br> <li> Implement <code>reconcileClusterConfigConfigMap()</code> method to create/update <br>ConfigMap with merged values<br> <li> Add reconciliation call in main <code>Reconcile()</code> method</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-632df3e422d02f42ebc3adf96fe1683930511f632b57b7d58fdcd60f75e46b2a">+92/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>main.go</strong><dd><code>Update reconciler initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/cmd/main.go

<ul><li>Update <code>KonfluxInfoReconciler</code> initialization to use new <br><code>NewKonfluxInfoReconciler()</code> constructor<br> <li> Replace struct literal initialization with constructor call</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-6951a7b3b8a59603497b68b62013a806397c640ce0a42a3862f6a4ffad4fda89">+5/-5</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>konfluxinfo_types_test.go</strong><dd><code>Unit tests for ClusterConfigData methods</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/api/v1alpha1/konfluxinfo_types_test.go

<ul><li>Add comprehensive unit tests for <code>ClusterConfigData.All()</code> iterator <br>method<br> <li> Add unit tests for <code>ClusterConfigData.MergeOver()</code> merge logic with <br>various scenarios<br> <li> Test empty fields, partial values, full values, and merge precedence<br> <li> Test early termination and field ordering in iterator</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-3661eb0af3d76fd505c9e6da3dc1a8d9b38cf3341cdc160c280f2e147df81103">+267/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>konfluxinfo_customization_test.go</strong><dd><code>Integration tests for cluster-config reconciliation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/internal/controller/info/konfluxinfo_customization_test.go

<ul><li>Add 7 integration tests for cluster-config ConfigMap creation and <br>updates<br> <li> Test user-provided values, empty values, partial values, and value <br>updates<br> <li> Test RBAC Role includes <code>cluster-config</code> in resourceNames<br> <li> Test merge behavior with injected <code>testClusterConfigDiscoverer</code> <br>implementation<br> <li> Add <code>testClusterConfigDiscoverer</code> helper for testing discovery <br>functionality</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-18a304ca14c8587ffae77de02bd3f1bc89b0aa27f00167c211bff5282b9a775d">+384/-0</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>manifests.yaml</strong><dd><code>Update RBAC for cluster-config access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/pkg/manifests/info/manifests.yaml

<ul><li>Add <code>cluster-config</code> to RBAC Role resourceNames for ConfigMap access</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-f4e9dfc60ef5ced8d5fdd269b5293dbf23e3a24aaeb87e0def4d2a0cd461d18f">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>rbac.yaml</strong><dd><code>Update RBAC for cluster-config access</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

operator/upstream-kustomizations/info/base/rbac.yaml

<ul><li>Add <code>cluster-config</code> to RBAC Role resourceNames for ConfigMap access</ul>

</details>

  </td>
  <td><a href="https://github.com/konflux-ci/konflux-ci/pull/4978/files#diff-d5cf85546e1da07156691254e95ced53d5510685b2462c34a27dd8fed082f399">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tbody></table>

</details>

___

### CODE REVIEW

File: operator/cmd/main.go [hunk 1]
```diff
@@ -359,11 +359,11 @@ func main() {
...
-	if err = (&info.KonfluxInfoReconciler{
-		Client:      mgr.GetClient(),
-		Scheme:      mgr.GetScheme(),
-		ObjectStore: objectStore,
-	}).SetupWithManager(mgr); err != nil {
+	if err = info.NewKonfluxInfoReconciler(
+		mgr.GetClient(),
+		mgr.GetScheme(),
+		objectStore,
+	).SetupWithManager(mgr); err != nil {
```
Comment (yftacherzog) [inline operator/cmd/main.go]: What's the reason the init pattern is now different here?

Comment (gbenhaim) [reply operator/cmd/main.go]: because I needed to create a an instance of `defaultClusterConfigDiscoverer`
I can remove the NewKonfluxInfoReconciler function and create the `defaultClusterConfigDiscoverer` in main if you find it better.

Comment (yftacherzog) [reply operator/cmd/main.go]: Was mostly curios about the change of pattern.

Comment (gbenhaim) [reply operator/cmd/main.go]: I'm going to  remove it. It doesn't bring any value ATM.
